ProjFS.Mac/PrjFSLib/PrjFSKextLogDaemon/PrjFSKextLogDaemon.cpp

```diff
@@ -0,0 +1,163 @@
+             if (connectResult != kIOReturnSuccess || connection == IO_OBJECT_NULL)
+             {
+                io_string_t servicePath = "";
+                IORegistryEntryGetPath(service, kIOServicePlane, servicePath);
+                if (connectResult == kIOReturnUnsupported)
+                {
+                    CFTypeRef kextVersionObj = IORegistryEntryCreateCFProperty(service, CFSTR(PrjFSKextVersionKey), kCFAllocatorDefault, 0);
+                    os_log_error(
+                        s_daemonLogger,
+                        "Failed to connect to newly matched PrjFS kernel service at '%{public}s'; version mismatch. Expected %{public}s, kernel service version %{public}@",
+                        servicePath, PrjFSKextVersion, kextVersionObj);
+                    if (kextVersionObj != nullptr)
+                    {
+                        CFRelease(kextVersionObj);
+                    }
+                }
+                else
+                {
+                    os_log_error(
+                        s_daemonLogger,
+                        "Failed to connect to newly matched PrjFS kernel service at '%{public}s'; connecting failed with error 0x%x",
+                        servicePath, connectResult);
+                }
+             }
+             else
+             {
+                 StartLoggingKextMessages(connection, service, s_daemonLogger, s_kextLogger);
+             }
+         });
+    if (nullptr == watchContext)
+    {
+        os_log_error(s_daemonLogger, "Failed to register for service notifications.");
+        return 1;
+    }
+    
+    SetupExitSignalHandler();
+
+    os_log(s_daemonLogger, "PrjFSKextLogDaemon running");
+
+    CFRunLoopRun();
+
+    os_log(s_daemonLogger, "PrjFSKextLogDaemon shutting down");
+
+    PrjFSService_StopWatching(watchContext);
+    
+    return 0;
+}
+
+static os_log_type_t KextLogLevelAsOSLogType(KextLog_Level level)
+{
+    switch (level)
+    {
+    case KEXTLOG_INFO:
+        return OS_LOG_TYPE_INFO;
+    case KEXTLOG_NOTE:
+        return OS_LOG_TYPE_DEFAULT;
+    case KEXTLOG_ERROR:
+    default:
+        return OS_LOG_TYPE_ERROR;
+    }
+}
+
+static void StartLoggingKextMessages(io_connect_t connection, io_service_t prjfsService, os_log_t daemonLogger, os_log_t kextLogger)
+{
+    uint64_t prjfsServiceEntryID = 0;
+    IORegistryEntryGetRegistryEntryID(prjfsService, &prjfsServiceEntryID);
+
+    std::shared_ptr<DataQueueResources> logDataQueue(new DataQueueResources {});
+    if (!PrjFSService_DataQueueInit(logDataQueue.get(), connection, LogPortType_MessageQueue, LogMemoryType_MessageQueue, dispatch_get_main_queue()))
+    {
+        os_log_error(s_daemonLogger, "Failed to set up log message data queue an connection to service with registry entry 0x%llx", prjfsServiceEntryID);
+        IOServiceClose(connection);
+        return;
+    }
+
+    os_log(s_daemonLogger, "Started logging messages from service with registry entry 0x%llx", prjfsServiceEntryID);
```

Comment:
Minor: We have a different component of VFS4G that we refer to as our "service", and so to avoid confusion it might be clearer to say `logging messages from kernel` or `from kext`.  

If "service" is more accurate, then switch from `from service` to something like`from PrjFS service`.

Reply:
I think something like `"Started logging messages from PrjFS IOService with registry entry id 0x%llx"` would be fully accurate and also unambiguous.